File: 0721-accounts-merge/0721-accounts-merge.py
# TC- O(eloge)
# SC- O(e) where e is total num of unique emails

class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        
        # connected components problem
        # each email is a vertex and is connected?

        # one name as key and multiple lists as values- this doesnt work if existing lists contain one matching element from new list- so you need to merge existing lists- so connecteed components approach is better

        emails_to_name = {}
        visited = set()
        adj_mat = defaultdict(list)

        for account in accounts:
            name = account[0]
            first_email = account[1]

            for email in account[1:]:
                emails_to_name[email] = name
                adj_mat[email].append(first_email)
                adj_mat[first_email].append(email)

        def dfs(email, component):
            visited.add(email)
            component.append(email)

            for neigh in adj_mat[email]:
                if neigh not in visited:
                    dfs(neigh, component)

        result = []
        for email in emails_to_name:
            if email not in visited:
                component = []
                dfs(email, component)
                result.append([emails_to_name[email]]+sorted(component))
        return result

File: 0721-accounts-merge/0721-accounts-merge.cpp
// TC- O(nlogn) where n is num of unique emails
// SC- O(n)

class Solution {
public:
    void dfs(const string& email, unordered_set<string>& visited, unordered_map<string, vector<string>>& adj_mat, vector<string>& connected){
            visited.insert(email);
            connected.push_back(email);

            for(string& neigh:adj_mat[email]){
                if(visited.find(neigh)==visited.end())
                    dfs(neigh, visited, adj_mat, connected);
            }
        }

    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        unordered_map<string, string> emailstoname;
        unordered_map<string, vector<string>> adjmat;

        for(auto& account:accounts){
            string name = account[0];
            string firstemail = account[1];
            
            for(int i=1; i<account.size(); i++){
                string email = account[i];
                emailstoname[email] = name;
                adjmat[email].push_back(firstemail);
                adjmat[firstemail].push_back(email);
            }
        }

        vector<vector<string>> res;
        unordered_set<string> visited;
        for(auto& [email, name]:emailstoname){
            if(visited.find(email)==visited.end()){
                vector<string> connected;
                dfs(email, visited, adjmat, connected);
                sort(connected.begin(), connected.end());
                connected.insert(connected.begin(), emailstoname[email]);
                res.push_back(connected);
            }
        }
        return res;
    }
};